File: tests/py/test_frame_format_deserialization.py
#!/usr/bin/env python3
"""
Frame format deserialization test for Python.

This test reads binary files and deserializes using different frame formats.
Usage: test_frame_format_deserialization.py <frame_format> <binary_file>
"""

import sys
import os
import json


def print_failure_details(label, expected_values=None, actual_values=None, raw_data=None):
    """Print detailed failure information"""
    print(f"\n{'='*60}")
    print(f"FAILURE DETAILS: {label}")
    print(f"{'='*60}")

    if expected_values:
        print("\nExpected Values:")
        for key, val in expected_values.items():
            print(f"  {key}: {val}")

    if actual_values:
        print("\nActual Values:")
        for key, val in actual_values.items():
            print(f"  {key}: {val}")

    if raw_data:
        print(f"\nRaw Data ({len(raw_data)} bytes):")
        print(f"  Hex: {raw_data.hex()}")

    print(f"{'='*60}\n")


def load_expected_values():
    """Load expected values from JSON file"""
    json_path = os.path.join(os.path.dirname(__file__), '..', 'expected_values.json')
    try:
        with open(json_path, 'r') as f:
            data = json.load(f)
        return data['serialization_test']
    except Exception as e:
        print(f"Error loading expected values: {e}")
        return None


def validate_message(msg, expected):
    """Validate a decoded message against expected values"""
    if msg.magic_number != expected['magic_number']:
        print_failure_details(
            "Value mismatch: magic_number",
            expected_values={"magic_number": expected['magic_number']},
            actual_values={"magic_number": msg.magic_number}
        )
        return False

    # Compare string (bytes vs string)
    test_string = msg.test_string.decode('utf-8').rstrip('\x00') if isinstance(
        msg.test_string, bytes) else str(msg.test_string).rstrip('\x00')
    if test_string != expected['test_string']:
        print_failure_details(
            "Value mismatch: test_string",
            expected_values={"test_string": expected['test_string']},
            actual_values={"test_string": test_string}
        )
        return False

    # Compare float with tolerance
    if abs(msg.test_float - expected['test_float']) > 0.0001:
        print_failure_details(
            "Value mismatch: test_float",
            expected_values={"test_float": expected['test_float']},
            actual_values={"test_float": msg.test_float}
        )
        return False

    if msg.test_bool != expected['test_bool']:
        print_failure_details(
            "Value mismatch: test_bool",
            expected_values={"test_bool": expected['test_bool']},
            actual_values={"test_bool": msg.test_bool}
        )
        return False

    # Compare array
    if list(msg.test_array) != expected['test_array']:
        print_failure_details(
            "Value mismatch: test_array",
            expected_values={"test_array": expected['test_array']},
            actual_values={"test_array": list(msg.test_array)}
        )
        return False

    return True


def get_frame_parser_class(frame_format):
    """Get the frame parser class for a given frame format name"""
    import importlib
    
    # Map frame format names to class names
    format_class_map = {
        'basic_default': 'BasicDefault',
        'basic_minimal': 'BasicMinimal',
        'basic_seq': 'BasicSeq',
        'basic_sys_comp': 'BasicSysComp',
        'basic_extended_msg_ids': 'BasicExtendedMsgIds',
        'basic_extended_length': 'BasicExtendedLength',
        'basic_extended': 'BasicExtended',
        'basic_multi_system_stream': 'BasicMultiSystemStream',
        'basic_extended_multi_system_stream': 'BasicExtendedMultiSystemStream',
        'tiny_default': 'TinyDefault',
        'tiny_minimal': 'TinyMinimal',
        'tiny_seq': 'TinySeq',
        'tiny_sys_comp': 'TinySysComp',
        'tiny_extended_msg_ids': 'TinyExtendedMsgIds',
        'tiny_extended_length': 'TinyExtendedLength',
        'tiny_extended': 'TinyExtended',
        'tiny_multi_system_stream': 'TinyMultiSystemStream',
        'tiny_extended_multi_system_stream': 'TinyExtendedMultiSystemStream',
    }
    
    class_name = format_class_map.get(frame_format)
    if not class_name:
        return None, None
    
    # Import the module dynamically using importlib
    # The generated code uses relative imports, so import through the package
    # The PYTHONPATH should be set to tests/generated (not tests/generated/py)
    try:
        # First try importing from py package
        module = importlib.import_module(f"py.{frame_format}")
        parser_class = getattr(module, class_name)
        return parser_class, class_name
    except (ImportError, AttributeError):
        pass
    
    # Fallback: try direct import (if PYTHONPATH includes generated/py directly)
    try:
        module = importlib.import_module(frame_format)
        parser_class = getattr(module, class_name)
        return parser_class, class_name
    except (ImportError, AttributeError) as e:
        print(f"  Error loading frame format {frame_format}: {e}")
        return None, None


def read_and_validate_test_data(frame_format, filename):
    """Read and validate test data from a binary file"""
    try:
        if not os.path.exists(filename):
            print(f"  Error: file not found: {filename}")
            return False

        with open(filename, 'rb') as f:
            binary_data = f.read()

        if len(binary_data) == 0:
            print_failure_details(
                "Empty file",
                expected_values={"data_size": ">0"},
                actual_values={"data_size": 0},
                raw_data=binary_data
            )
            return False

        # Import the message class (works with PYTHONPATH including generated parent)
        import importlib
        try:
            msg_module = importlib.import_module('py.serialization_test_sf')
            SerializationTestSerializationTestMessage = msg_module.SerializationTestSerializationTestMessage
        except ImportError:
            try:
                from serialization_test_sf import SerializationTestSerializationTestMessage
            except ImportError as e:
                print(f"  Error importing serialization_test_sf: {e}")
                return False

        # Get frame parser class
        parser_class, class_name = get_frame_parser_class(frame_format)
        if parser_class is None:
            print(f"  Unknown frame format: {frame_format}")
            return False

        # Validate and decode using the frame parser
        result = parser_class.validate_packet(list(binary_data))
        
        if not result.valid:
            print_failure_details(
                "Failed to decode data",
                expected_values={"decoded_message": "valid"},
                actual_values={"decoded_message": None},
                raw_data=binary_data
            )
            return False

        # Decode the message data
        msg_data = bytes(result.msg_data)
        decoded_msg = SerializationTestSerializationTestMessage.create_unpack(msg_data)

        # Load expected values and validate
        expected = load_expected_values()
        if not expected:
            return False

        if not validate_message(decoded_msg, expected):
            print("  Validation failed")
            return False

        print(f"  [OK] Data validated successfully with {frame_format} format")
        return True

    except ImportError as e:
        print(f"  Error: Generated code not available: {e}")
        return False

    except Exception as e:
        print_failure_details(
            f"Read data exception: {type(e).__name__}",
            expected_values={"result": "success"},
            actual_values={"exception": str(e)}
        )
        import traceback
        traceback.print_exc()
        return False


def get_supported_formats():
    """Return list of supported frame formats"""
    return [
        'basic_default', 'basic_minimal', 'basic_seq', 'basic_sys_comp',
        'basic_extended_msg_ids', 'basic_extended_length', 'basic_extended',
        'basic_multi_system_stream', 'basic_extended_multi_system_stream',
        'tiny_default', 'tiny_minimal', 'tiny_seq', 'tiny_sys_comp',
        'tiny_extended_msg_ids', 'tiny_extended_length', 'tiny_extended',
        'tiny_multi_system_stream', 'tiny_extended_multi_system_stream',
    ]


def main():
    """Main test function"""
    print("\n[TEST START] Python Frame Format Deserialization")

    if len(sys.argv) != 3:
        print(f"  Usage: {sys.argv[0]} <frame_format> <binary_file>")
        print("  Supported formats:")
        for fmt in get_supported_formats():
            print(f"    {fmt}")
        print("[TEST END] Python Frame Format Deserialization: FAIL\n")
        return False

    frame_format = sys.argv[1]
    filename = sys.argv[2]
    
    success = read_and_validate_test_data(frame_format, filename)

    status = "PASS" if success else "FAIL"
    print(f"[TEST END] Python Frame Format Deserialization: {status}\n")

    return success


if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)

File: tests/ts/test_frame_format_deserialization.ts
/**
 * Frame format deserialization test for TypeScript.
 * 
 * This test reads binary files and deserializes using different frame formats.
 * Usage: test_frame_format_deserialization.ts <frame_format> <binary_file>
 */
import * as fs from 'fs';
import * as path from 'path';

function printFailureDetails(label: string): void {
  console.log('\n============================================================');
  console.log(`FAILURE DETAILS: ${label}`);
  console.log('============================================================\n');
}

function loadExpectedValues(): any {
  try {
    const jsonPath = path.join(__dirname, '../../../expected_values.json');
    const data = JSON.parse(fs.readFileSync(jsonPath, 'utf-8'));
    return data.serialization_test;
  } catch (error) {
    console.log(`Error loading expected values: ${error}`);
    return null;
  }
}

interface FrameFormatConfig {
  start: number[];
  headerSize: number;
  footerSize: number;
  hasLength: boolean;
  lengthBytes: number;
}

// Frame format configurations
const FRAME_FORMATS: { [key: string]: FrameFormatConfig } = {
  'basic_default': { start: [0x90, 0x71], headerSize: 4, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'basic_minimal': { start: [0x90, 0x70], headerSize: 3, footerSize: 0, hasLength: false, lengthBytes: 0 },
  'basic_seq': { start: [0x90, 0x76], headerSize: 5, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'basic_sys_comp': { start: [0x90, 0x75], headerSize: 6, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'basic_extended_msg_ids': { start: [0x90, 0x72], headerSize: 5, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'basic_extended_length': { start: [0x90, 0x73], headerSize: 5, footerSize: 2, hasLength: true, lengthBytes: 2 },
  'basic_extended': { start: [0x90, 0x74], headerSize: 6, footerSize: 2, hasLength: true, lengthBytes: 2 },
  'basic_multi_system_stream': { start: [0x90, 0x77], headerSize: 7, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'basic_extended_multi_system_stream': { start: [0x90, 0x78], headerSize: 9, footerSize: 2, hasLength: true, lengthBytes: 2 },
  'tiny_default': { start: [0x71], headerSize: 3, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'tiny_minimal': { start: [0x70], headerSize: 2, footerSize: 0, hasLength: false, lengthBytes: 0 },
  'tiny_seq': { start: [0x76], headerSize: 4, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'tiny_sys_comp': { start: [0x75], headerSize: 5, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'tiny_extended_msg_ids': { start: [0x72], headerSize: 4, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'tiny_extended_length': { start: [0x73], headerSize: 4, footerSize: 2, hasLength: true, lengthBytes: 2 },
  'tiny_extended': { start: [0x74], headerSize: 5, footerSize: 2, hasLength: true, lengthBytes: 2 },
  'tiny_multi_system_stream': { start: [0x77], headerSize: 6, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'tiny_extended_multi_system_stream': { start: [0x78], headerSize: 8, footerSize: 2, hasLength: true, lengthBytes: 2 },
};

function validateFrameFormat(buffer: Buffer, frameFormat: string, expected: any): boolean {
  const config = FRAME_FORMATS[frameFormat];
  if (!config) {
    console.log(`  Unknown frame format: ${frameFormat}`);
    return false;
  }

  // Check minimum length
  if (buffer.length < config.headerSize + config.footerSize) {
    console.log(`  Data too short for ${frameFormat} format`);
    return false;
  }

  // Check start bytes
  for (let i = 0; i < config.start.length; i++) {
    if (buffer[i] !== config.start[i]) {
      console.log(`  Invalid start byte at position ${i} (expected 0x${config.start[i].toString(16)}, got 0x${buffer[i].toString(16)})`);
      return false;
    }
  }

  // Validate CRC if present
  if (config.footerSize > 0) {
    let byte1 = 0;
    let byte2 = 0;
    const checksumStart = config.start.length;
    const checksumEnd = buffer.length - config.footerSize;
    
    for (let i = checksumStart; i < checksumEnd; i++) {
      byte1 = (byte1 + buffer[i]) & 0xFF;
      byte2 = (byte2 + byte1) & 0xFF;
    }
    
    if (byte1 !== buffer[buffer.length - 2] || byte2 !== buffer[buffer.length - 1]) {
      console.log(`  Checksum mismatch`);
      return false;
    }
  }

  // Extract magic number from payload
  const payloadStart = config.headerSize;
  const magicNumber = buffer.readUInt32LE(payloadStart);
  if (magicNumber !== expected.magic_number) {
    console.log(`  Magic number mismatch (expected ${expected.magic_number}, got ${magicNumber})`);
    return false;
  }

  console.log(`  [OK] Data validated successfully with ${frameFormat} format`);
  return true;
}

function readAndValidateTestData(frameFormat: string, filename: string): boolean {
  try {
    if (!fs.existsSync(filename)) {
      console.log(`  Error: file not found: ${filename}`);
      return false;
    }

    const binaryData = fs.readFileSync(filename);

    if (binaryData.length === 0) {
      printFailureDetails('Empty file');
      return false;
    }

    const expected = loadExpectedValues();
    if (!expected) {
      return false;
    }

    if (!validateFrameFormat(binaryData, frameFormat, expected)) {
      console.log('  Validation failed');
      return false;
    }

    return true;
  } catch (error) {
    printFailureDetails(`Read data exception: ${error}`);
    return false;
  }
}

function main(): boolean {
  console.log('\n[TEST START] TypeScript Frame Format Deserialization');

  const args = process.argv.slice(2);
  if (args.length !== 2) {
    console.log(`  Usage: ${process.argv[1]} <frame_format> <binary_file>`);
    console.log('  Supported formats:');
    for (const fmt of Object.keys(FRAME_FORMATS)) {
      console.log(`    ${fmt}`);
    }
    console.log('[TEST END] TypeScript Frame Format Deserialization: FAIL\n');
    return false;
  }

  try {
    const success = readAndValidateTestData(args[0], args[1]);

    console.log(`[TEST END] TypeScript Frame Format Deserialization: ${success ? 'PASS' : 'FAIL'}\n`);
    return success;
  } catch (error) {
    printFailureDetails(`Exception: ${error}`);
    console.log('[TEST END] TypeScript Frame Format Deserialization: FAIL\n');
    return false;
  }
}

if (require.main === module) {
  const success = main();
  process.exit(success ? 0 : 1);
}

export { main };
